### ISSUE TITLE
[Go][Parquet] Trouble reading a Parquet file with a RecordReader

### ISSUE THREAD

**Issue by tschaub** [apache/arrow#37402]
### Describe the usage question you have. Please include as many useful details as  possible.

I've seen a suggestion on Stack Overflow by @zeroshade to use [`pqarrow.RecordReader`](https://pkg.go.dev/github.com/apache/arrow/go/v13@v13.0.0/parquet/pqarrow#RecordReader) to simplify reading records with repeated or nested values.  I'm looking at [the tests](https://github.com/apache/arrow/blob/go/v13.0.0/go/parquet/pqarrow/file_reader_test.go#L147) for examples, but struggling with some basic usage.

It looks like I can read all records if I use a batch size of 1.  For example, this test passes:
```go
func TestRecordReading(t *testing.T) {
	f, fileErr := os.Open("testdata/example.parquet")
	require.NoError(t, fileErr)
	reader, readerErr := file.NewParquetReader(f)
	require.NoError(t, readerErr)
	defer reader.Close()

	pqReader, pqErr := pqarrow.NewFileReader(reader, pqarrow.ArrowReadProperties{BatchSize: 1}, memory.DefaultAllocator)
	require.NoError(t, pqErr)

	recordReader, rrErr := pqReader.GetRecordReader(context.Background(), nil, nil)
	require.NoError(t, rrErr)

	numRows := 0
	for {
		rec, err := recordReader.Read()
		if err == io.EOF {
			assert.Nil(t, rec)
			break
		}
		assert.NoError(t, err)
		numRows += 1
	}

	assert.Equal(t, reader.NumRows(), int64(numRows))
}
```

However, if I instead use `pqarrow.ArrowReadProperties{BatchSize: 2}` or any other batch size, I get `io.EOF` before all records are read.

Is this the right pattern for reading records from a Parquet file?  If does the total number of records in the file need to be an integer multiple of the batch size used in order to read all records?

### Component(s)

Go, Parquet

**Comment by mapleFU** [comment]
In C++ and Rust implementions, `RecordReader` would work if `numRows % batchSize != 0`. I'm not sure it's aligned, if it's, maybe I can draft a fixing for this

**Comment by tschaub** [comment]
Thanks for the reply and for putting together a fix, @mapleFU.

**Comment by tschaub** [comment]
I see now that I might have misinterpreted a "record" as corresponding to a single row.  It looks like the test above passes if I increment the `numRows` by `rec.NumRows()` instead of by 1 (I was assuming each record represented a single row).

```diff
 func TestRecordReading(t *testing.T) {
        f, fileErr := os.Open("testdata/example.parquet")
        require.NoError(t, fileErr)
        reader, readerErr := file.NewParquetReader(f)
        require.NoError(t, readerErr)
        defer reader.Close()
 
-       pqReader, pqErr := pqarrow.NewFileReader(reader, pqarrow.ArrowReadProperties{BatchSize: 1}, memory.DefaultAllocator)
+       pqReader, pqErr := pqarrow.NewFileReader(reader, pqarrow.ArrowReadProperties{BatchSize: 10}, memory.DefaultAllocator)
        require.NoError(t, pqErr)
 
        recordReader, rrErr := pqReader.GetRecordReader(context.Background(), nil, nil)
        require.NoError(t, rrErr)
 
        numRows := 0
        for {
                rec, err := recordReader.Read()
                if err == io.EOF {
                        assert.Nil(t, rec)
                        break
                }
                assert.NoError(t, err)
-               numRows += 1
+               numRows += int(rec.NumRows())
        }
 
        assert.Equal(t, reader.NumRows(), int64(numRows))
 }
```

I thought the batch size had to do with the size of a buffer used internally when reading column values, and didn't understand that it would determine the number of rows per record from a record reader.

**Comment by mapleFU** [comment]
I dont know why test passed. Would you mind update `example.parquet`? It would helps a lot.

After check, I found my fault, because Go Parquet would handle unaligned `numRows` well. However, it cannot handle empty RowGroup. So: if row-group is not empty, I guess it would works well in most cases, however if row-group is empty, it would cause error.

**Comment by tschaub** [comment]
@mapleFU - In my case the `example.parquet` only has a single row group with 5 rows in it - so not very representative, just data created for the tests.  I'll try to generate some additional data to see where things might not work.

I'm also seeing panics using a RecordReader on data with some null byte arrays (and `DELTA_LENGTH_BYTE_ARRAY` encoding).  Trying to narrow things down to an easily reproducible test case.

**Comment by mapleFU** [comment]
It might because https://github.com/apache/arrow/issues/37102 . It's checked in after 13.0 freeze

**Comment by tschaub** [comment]
I've attached an archive with three test files.  The `example.parquet` file passes the test above (it only has 5 rows in a single row group).

The same test panics with either the `sparse-byte-arrays.parquet` or `sparse-strings.parquet`.  I've included detail on the panics here: https://gist.github.com/tschaub/adfa736459398770b5cff327dc81ffe1.  The `sparse-strings.parquet` panic looks similar to #37102.  I'll try running both of these using your fix from #37112.

[testdata.zip](https://github.com/apache/arrow/files/12450523/testdata.zip)

Thanks for your help with this, @mapleFU.
